token-lending/program/src/processor.rs

```diff
@@ -578,6 +587,309 @@ fn _deposit_reserve_liquidity<'a>(
+        lending_market_authority_info,
+        sysvar_info,
+        token_program_id,
+    )?;
+    Ok(())
+}
+
+#[allow(clippy::too_many_arguments)]
+fn _flash_borrow_reserve_liquidity<'a>(
+    program_id: &Pubkey,
+    liquidity_amount: u64,
+    source_liquidity_info: &AccountInfo<'a>,
+    destination_liquidity_info: &AccountInfo<'a>,
+    reserve_info: &AccountInfo<'a>,
+    lending_market_info: &AccountInfo<'a>,
+    lending_market_authority_info: &AccountInfo<'a>,
+    sysvar_info: &AccountInfo<'a>,
+    token_program_id: &AccountInfo<'a>,
+) -> Result<u64, ProgramError> {
+    let lending_market = LendingMarket::unpack(&lending_market_info.data.borrow())?;
+    if lending_market_info.owner != program_id {
+        msg!("Lending market provided is not owned by the lending program");
+        return Err(LendingError::InvalidAccountOwner.into());
+    }
+    if &lending_market.token_program_id != token_program_id.key {
+        msg!("Lending market token program does not match the token program provided");
+        return Err(LendingError::InvalidTokenProgram.into());
+    }
+    let mut reserve = Reserve::unpack(&reserve_info.data.borrow())?;
+    if reserve_info.owner != program_id {
+        msg!("Reserve provided is not owned by the lending program");
+        return Err(LendingError::InvalidAccountOwner.into());
+    }
+    if &reserve.lending_market != lending_market_info.key {
+        msg!("Reserve lending market does not match the lending market provided");
+        return Err(LendingError::InvalidAccountInput.into());
+    }
+    if &reserve.liquidity.supply_pubkey != source_liquidity_info.key {
+        msg!("Borrow reserve liquidity supply must be used as the source liquidity provided");
+        return Err(LendingError::InvalidAccountInput.into());
+    }
+    if &reserve.liquidity.supply_pubkey == destination_liquidity_info.key {
+        msg!(
+            "Borrow reserve liquidity supply cannot be used as the destination liquidity provided"
+        );
+        return Err(LendingError::InvalidAccountInput.into());
+    }
+    if reserve.borrowing {
+        msg!("Multiple flash borrows not allowed");
+        return Err(LendingError::MultipleFlashBorrows.into());
+    }
+
+    let authority_signer_seeds = &[
+        lending_market_info.key.as_ref(),
+        &[lending_market.bump_seed],
+    ];
+    let lending_market_authority_pubkey =
+        Pubkey::create_program_address(authority_signer_seeds, program_id)?;
+    if &lending_market_authority_pubkey != lending_market_authority_info.key {
+        msg!(
+            "Derived lending market authority {} does not match the lending market authority provided {}",
+            &lending_market_authority_pubkey.to_string(),
+            &lending_market_authority_info.key.to_string(),
+        );
+        return Err(LendingError::InvalidMarketAuthority.into());
+    }
+
+    // @FIXME: if u64::MAX is flash loaned, fees should be inclusive as with ordinary borrows
```

Comment:
This comment was copied over from the existing flash loan implementation in this file, but I am not sure exactly what it means.

Reply:
i think there is perhaps an overflow issue? like if you borrow and borrow +fees > u64::MAX maybe need to handle that differently?

Reply:
I think you're correct. If we lent `u64::MAX`, then the tx would fail with `MathOverflow` due to a `checked_add`. So the only way to support this would be to include the fees in the calculation. Therefore user would be lent `u64::MAX - fees` and we would expect `u64::MAX` to be returned.